Support multiline textfields for printing

calixteman opened the pull request.

timvandermeij commented on src/core/annotation.js:
```diff
@@ -1081,6 +1094,92 @@ class WidgetAnnotation extends Annotation {
+  handleMultiline(text, font, fontSize, width, alignment, hPadding, vPadding) {
+    const lines = text.replace("\r\n", "\n").split("\n");
+    let buf = "";
+    const totalWidth = alignment === 1 ? width : width - 2 * hPadding;
```
This could use a comment. What is `alignment === 1`; center, left of right? Why is that special cased?

timvandermeij commented on src/core/annotation.js:
```diff
@@ -1081,6 +1094,92 @@ class WidgetAnnotation extends Annotation {
+            width,
+            alignment,
+            hPadding,
+            -fontSize
```
Why take the negative here?

calixteman replied:
I added a comment to explain.

timvandermeij commented on src/core/annotation.js:
```diff
@@ -1081,6 +1094,92 @@ class WidgetAnnotation extends Annotation {
+    return buf;
+  }
+
+  splitLine(line, font, fontSize, width) {
```
Let's rename this to `_splitLine` to make it private.

timvandermeij commented on src/core/annotation.js:
```diff
@@ -1081,6 +1094,92 @@ class WidgetAnnotation extends Annotation {
+
+  splitLine(line, font, fontSize, width) {
+    const scale = fontSize / 1000;
+    const white = font.charsToGlyphs(" ", true)[0].width * scale;
```
I think `const whitespace` would be bit more descriptive here.

timvandermeij commented on src/core/annotation.js:
```diff
@@ -1081,6 +1094,92 @@ class WidgetAnnotation extends Annotation {
+      startChunk = 0,
+      currentWidth = 0;
+
+    for (let i = 0; i < line.length; i++) {
```
We usually do `for (let i = 0, ii = line.length; i < ii; i++) {` to cache the line length once.

timvandermeij commented on src/core/annotation.js:
```diff
@@ -1081,6 +1094,92 @@ class WidgetAnnotation extends Annotation {
+      } else {
+        const charWidth = font.charsToGlyphs(character, false)[0].width * scale;
+        if (currentWidth + charWidth > width) {
+          // We must break to the last white position (if one)
```
`last whitespace position (if available)`

timvandermeij commented on test/unit/annotation_spec.js:
```diff
@@ -1638,6 +1638,41 @@ describe("annotation", function () {
         }, done.fail);
     });
+
+    it("should render splited text for printing", function (done) {
```
`should render multiline text for printing`

timvandermeij commented on test/unit/annotation_spec.js:
```diff
@@ -1638,6 +1638,41 @@ describe("annotation", function () {
+        .then(annotation => {
+          const id = annotation.data.id;
+          const annotationStorage = {};
+          annotationStorage[id] = "a aa aaa aaaa aaaaa aaaaaa";
```
Can we make this string so that it also covers the "break in the middle of string" (no whitespace found) codepath, since that seems to be uncovered here? Perhaps introducing a large word would already be enough. If this line becomes longer than 80 characters, then concatenate smaller substrings.

timvandermeij commented on test/unit/annotation_spec.js:
```diff
@@ -1638,6 +1638,41 @@ describe("annotation", function () {
+        }, done.fail)
+        .then(appearance => {
+          expect(appearance).toEqual(
+            "/Tx BMC q BT /Helv 5 Tf 1 0 0 1 0 10 Tm 2.00 -5.00 Td (a aa aaa ) Tj\n0.00 -5.00 Td (aaaa aaaaa ) Tj\n0.00 -5.00 Td (aaaaaa) Tj ET Q EMC"
```
Same here about concatenating smaller substrings to stay within the 80 character limit.

timvandermeij reviewed (commented):
Looks good with the comments addressed (only the second commit was checked here since the rest is in another PR).

calixteman committed:
Support multiline textfields for printing

timvandermeij commented:
/botio test

timvandermeij commented:
/botio-linux preview

timvandermeij commented:
This doesn't seem to be working correctly. I tried inserting text into the multiline textfield from https://github.com/mozilla/pdf.js/blob/master/test/pdfs/annotation-text-widget.pdf (and other PDFs), but the print result is an empty textfield.

calixteman commented:
I didn't manage to reproduce using print button but I managed to have an incorrect behaviour using keyboard shortcut (and in this case `change` event was not triggered so I replaced it with `input` event).
Anyway, could you try again ?
I fixed a bug with `\r` as line separator and with non-empty fields which have been changed into empty ones.
I'll add some tests in the weekend.

timvandermeij commented:
/botio-linux preview

timvandermeij commented:
Unfortunately the problem remains. Steps to reproduce:

1. Visit http://54.67.70.0:8877/a931282262c09f9/web/viewer.html (preview link above)
2. Run `PDFViewerApplication.preferences.set('renderInteractiveForms', true)` in the console and refresh.
3. Download https://github.com/mozilla/pdf.js/blob/master/test/pdfs/annotation-text-widget.pdf and open the file with the viewer.
4. Enter multiline text in the textarea.
5. Click the print button in the toolbar or Ctrl-P and print the result to PDF.

Attached is the result I get from this: 
[out.pdf](https://github.com/mozilla/pdf.js/files/5045344/out.pdf)

timvandermeij commented on src/display/annotation_layer.js:
```diff
@@ -462,7 +462,7 @@ class TextWidgetAnnotationElement extends WidgetAnnotationElement {
       }
 
-      element.addEventListener("change", function (event) {
+      element.addEventListener("input", function (event) {
```
If this is a problem here, shouldn't it also be changed for checkbox and radio button widgets from previous PRs? If so, please change that in a separate commit.

calixteman commented:
It was because of the alignment var created after its use... I wonder why neither the linter nor firefox dev edition didn't catch it...
Anyway I fixed it and added a test for multiple possible EOL.
About the event: https://developer.mozilla.org/en-US/docs/Web/API/HTMLElement/input_event
So it seems that `change` is ok for checkbox and radios.

calixteman commented:
/botio-linux preview

timvandermeij commented:
/botio-linux preview

timvandermeij commented:
/botio test

timvandermeij commented:
This works fine now; thanks!
